doc/services/device_mgmt/mcumgr_handlers.rst

```diff
@@ -0,0 +1,516 @@
+responses that return a :c:enum:`mcumgr_err_t` - there should always be an OK error code with the
+value 0 and an unknown error code with the value 1. The above example then adds an error code of
+``not wanted`` with value 2. In addition, the group ID is set to be
+:c:enum:`MGMT_GROUP_ID_PERUSER`, which is the start group ID for user-defined groups, note that
+group IDs need to be unique so other custom groups should use different values, a central index
+header file (as upstream Zephyr has) can be used to distribute group IDs more easily.
+
+Initial header <grp_name>_mgmt_callbacks.h
+==========================================
+
+The purpose of the header file is to provide defines which can be used by the MCUmgr handler
+itself and application code, e.g. to reference the command IDs for executing functions. An example
+file would look similar to:
+
+.. code-block:: c
+
+    /*
+     * Copyright (c) 2023 Nordic Semiconductor ASA
+     *
+     * SPDX-License-Identifier: Apache-2.0
+     */
+
+    #ifndef H_MCUMGR_EXAMPLE_MGMT_CALLBACKS_
+    #define H_MCUMGR_EXAMPLE_MGMT_CALLBACKS_
+
+    #include <stdint.h>
+    #include <zephyr/mgmt/mcumgr/mgmt/callbacks.h>
+
+    #ifdef __cplusplus
+    extern "C" {
+    #endif
+
+    /* This is the event ID for the example group */
+    #define MGMT_EVT_GRP_EXAMPLE MGMT_EVT_GRP_USER_CUSTOM_START
+
+    /* MGMT event opcodes for example management group */
+    enum example_mgmt_group_events {
+            /* Callback when the other command is received, data is example_mgmt_other_data */
+            MGMT_EVT_OP_EXAMPLE_OTHER     = MGMT_DEF_EVT_OP_ID(MGMT_EVT_GRP_EXAMPLE, 0),
+
+            /* Used to enable all smp_group events */
+            MGMT_EVT_OP_EXAMPLE_MGMT_ALL  = MGMT_DEF_EVT_OP_ALL(MGMT_EVT_GRP_EXAMPLE),
+    };
+
+
+    /* Structure provided in the #MGMT_EVT_OP_EXAMPLE_OTHER notification callback */
+    struct example_mgmt_other_data {
+            /* Contains the user supplied value */
+            uint32_t user_value;
+    };
+
+    #ifdef __cplusplus
+    }
+    #endif
+
+    #endif
+
+This sets up a single event which application (or module) code can register for to receive a
+callback when the function handler is executed, which allows the flow of the handler to be
+changed (i.e. to return an error instead of continuing). The event group ID is set to
+:c:enum:`MGMT_EVT_GRP_USER_CUSTOM_START`, which is the start event ID for user-defined groups,
+note that event IDs need to be unique so other custom groups should use different values, a
+central index header file (as upstream Zephyr has) can be used to distribute event IDs more
+easily.
+
+Initial source <grp_name>_mgmt.c
+================================
+
+The purpose of this source file is to handle the incoming MCUmgr commands, provide responses, and
+register the transport with MCUmgr so that commands will be sent to it. An example file would
+look similar to:
+
+.. code-block:: c
+
```

Comment:
Wow that's a lot of code for a document. Thinking, it's going to be hard to keep this up to date, would it be an option to have these live in actual C files in samples or tests and have them build by the CI and included here instead? I think Go has a feature to build code for examples and that's useful to make sure stuff don't go stale, having so much code in document seems like asking for troubles.

@kartben @gmarull

Reply:
Thanks for the flag, @fabiobaltieri, and +1.
Is there a reason why this couldn't be an actual sample? Don't get me wrong, I do love the idea of inlining the code snippets in the doc, but this could be done using `.. literalinclude:: ...`, an under-utilized Sphinx feature IMO :) (and it can be pretty flexible too, i.e. only sections of a file can be include, etc.)

Reply:
+1 on using `literalinclude::` pointing to actual code.  The [USB-C power delivery documentation](https://github.com/zephyrproject-rtos/zephyr/blob/3a24476fb70b813cf5ba01853c10f5c7973dbfbc/doc/connectivity/usb/pd/ucds.rst?plain=1#L39) provides some good examples.

Reply:
The problem I have with making this a sample or a test is for a test, I don't want to test this, this functionality is tested already in other tests, to make all the code for testing this operation is an absolute time sink and doesn't provide anything. As for a sample, I don't really see the point, because no-one would actually use this code as-is, they would look at it to refer to how handlers are implemented but create their own (because to use this handler, you would need a corresponding MCUmgr client which supported it, which they would likely have to go and create themselves). So it can be made a sample if wanted but isn't going to be a useful sample. MCUmgr is a stable API so these APIs are not going to change unless the existing APIs are deprecated or a stable API change request is made

Reply:
I don't think the idea was tho have it runnable, the idea was to have CI checking that it builds as the rest of the ecosystem moves. But it'd also make it easier to use, one can just copy the files out rather than copypasting the content. How much work is to move these under `tests`, and add the support files to make it a build only test?

Reply:
Nice, that does work pretty well, have updated it